### ISSUE TITLE
DateTime String Interpolation Issue

### ISSUE THREAD

**Issue by andreweberle** [dotnet/efcore#33264]
**Title:** Issue with FormattableString Formatting in FromSqlInterpolated

EF Core: 8.0.2

**Description:**
When creating a `FormattableString` for a SQL query to be used with `FromSqlInterpolated` in Entity Framework Core, there seems to be an inconsistency in how the formatting is applied, resulting in unexpected behavior. 

**Issue:**
When constructing the `FormattableString` as shown below:

```csharp
return $@"SELECT * FROM [DB].[dbo].[Orders] 
         WHERE CONVERT(DATE, OrderDate AT TIME ZONE 'UTC' AT TIME ZONE {timeZoneInfo.Id}) 
         = {startDateTime:yyyy-MM-dd}";
```

The intention is to format the `startDateTime` variable in the specified format (`yyyy-MM-dd`). However, when this `FormattableString` is passed to `FromSqlInterpolated`, the formatting does not seem to be honored, resulting in a SQL query with parameters that are not in the expected format.

**Example:**
```sql
DECLARE p0 nvarchar(4000) = N'AUS Eastern Standard Time';
DECLARE p1 datetime2 = '2024-03-07T17:50:58.0365954';

SELECT * FROM [DB].[dbo].[Orders] 
WHERE CONVERT(DATE, OrderDate AT TIME ZONE 'UTC' AT TIME ZONE @p0) 
= @p1
```

Here, the `@p1` parameter is not in the expected format (`yyyy-MM-dd`), which can lead to unexpected results or errors when executing the query.

Resolution Attempt:
When manually formatting the `startDateTime` using `ToString("yyyy-MM-dd")`, the resulting SQL query parameters are in the correct format.

```csharp
$@"SELECT * FROM [DB.[dbo].[Orders] 
   WHERE CONVERT(DATE, OrderDate AT TIME ZONE 'UTC' AT TIME ZONE {timeZoneInfo.Id}) 
   = {startDateTime.ToString("yyyy-MM-dd")}";
```

```sql
DECLARE p0 nvarchar(4000) = N'AUS Eastern Standard Time';
DECLARE p1 datetime2 = '2024-03-07';

SELECT * FROM [DB].[dbo].[Orders] 
WHERE CONVERT(DATE, OrderDate AT TIME ZONE 'UTC' AT TIME ZONE @p0) 
= @p1
```

This suggests that there might be an issue with how `FormattableString` is being handled or interpreted by `FromSqlInterpolated`.

**Expected Behavior:**
The `FormattableString` should honor the specified formatting when used with `FromSqlInterpolated`, ensuring that the parameters in the resulting SQL query are in the expected format.

**Comment by ajcvickers** [comment]
Note for team triage: the request here is to use the formatting info from the interpolated string to format the parameter. As far as I am aware, this is never something we have considered, and given that in general parameterization for the database does not simply string formatting, I'm not sure it makes sense to do this. But we can discuss.

**Comment by roji** [comment]
When FormattableString is used in .NET to produce an actual string, formatting instructions (such as `{startDateTime:yyyy-MM-dd}`) make sense - the DateTime is formatted as a string with the given pattern, and the result is inserted as a simple string into the larger string.

However, when using FormattableString to specify a SQL command for execution against the database, the end result isn't a single string that contains the FormattableString's arguments - the arguments become type parameters which are sent outside the SQL itself (both for efficiency and to prevent SQL injection attacks). Because of this, I don't think it makes sense to apply formatting to your DateTime, since that would mean that the parameter is no longer a SQL `datetime2`, but rather e.g. an `nvarchar(max)`.

This is exactly what you do in your workaround (`{startDateTime.ToString("yyyy-MM-dd")}`) - EF sends a string parameters instead of a timestamp. Now, SQL Server may support implicitly converting the `nvarchar(max)` back to a `datetime2` based on the context in the SQL query, but this is problematic in many ways: it won't work for many types, it won't work for various other databases (e.g. PostgreSQL doesn't support such implicit conversions), and crucially, the extra conversion  may also prevent index usage, causing the query to run significantly slower than it should. If we allowed this, I believe many users would fall into this pit of failure, not fully understand that this isn't a conversion from DateTime to DateTime, but rather from DateTime to string (with all the problematic consequences), and so I think it's better to not do this.

To summarize, while I understand the expectation here, ultimately this is a bit of a mismatch when using a mechanism that was meant more for constructing strings (FormattableString) for doing something else (constructing a SQL command with parameters).

However, I do think we should at least throw in such case, rather than simply ignoring the formatting.

**Comment by roji** [comment]
Closing as everyone seems to agree that we shouldn't go in this direction - but feel free to post back here for further discussion.

**Comment by andreweberle** [comment]
@roji Thank you kindly for taking the time to look into this.
